### PR TITLE
Add copy button for code snippet

### LINKED ISSUE #2923: Add a code snippet copy button
**What would you like to be added**:

We are a programming blog platform and a lot of code is written every week. It would be nice to have a `Copy` button on the top right of every code block to encourage people to include their code snippets or terminal logs so others can copy to try it out or help with debugging.

### PR DESCRIPTION
**Pull request by menghif**
## Issue This PR Addresses
Fixes #2923 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

Image 2022-03-27 at 11 50 PM 2
Image 2022-03-27 at 11 50 PM

## Steps to test the PR

- `pnpm i`
- `cp config/env.staging .env`
- `pnpm dev`

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by humphd** [comment]
In our meeting today we discussed copying how GitHub does this, so that you only see the "copy" button when you mouse over the area, and it is removed when you leave that area.  We should see what they do with touch events for mobile.

Also, I wonder if we should steal their icon too vs. using the "copy" text?

File: src/web/src/pages/CopyButton.css [hunk 1]
```diff
@@ -0,0 +1,22 @@
+.copyCodeBtn {
+  position: absolute;
+  right: 0;
+  margin-right: 1.5rem;
+  padding: 0.4rem;
+  border: 1px solid #121d59;
+  border-radius: 0.5rem;
+  background-color: #ffffff;
+  color: #121d59;
+  transition-duration: 0.4s;
```
Comment (DukeManh) [inline src/web/src/pages/CopyButton.css:10]:
```suggestion
  transition-duration: 0.2s;
```
Keep it under 300ms for a snappier feel.

**Review by DukeManh — DISMISSED** [review]
Instead of listening to the mouse move, you can use `mouseenter/mouseleave` event handler. I suggest that you do the followings:

1. `const snippets = post.querySelectorAll('pre code')`
2. `snippet.onmouseenter = () => {
const button = createButton();
button.onclick = () => copySnippet(snippet.text);
snippet.appendChild(button);
}`
3. `snippet.onmouseleave =  removeButton`

**Comment by TueeNguyen** [comment]
@menghif, I was playing around with this PR, trying to make `mouseenter` & `mouseleave` to work. I opened a PR in your fork

File: src/web/src/components/Posts/Post.tsx [hunk 2]
```diff
@@ -360,9 +395,17 @@ const PostComponent = ({ postUrl, currentPost, totalPosts }: Props) => {
...
+      window.document.removeEventListener('click', handleClick);
+    };
+  }, []);
+
+  // Listen for mouse move events
+  useEffect(() => {
+    window.document.addEventListener('mousemove', handleMouseMove);
     return () => {
-      window.document.removeEventListener('click', handleZoom);
+      window.document.removeEventListener('mousemove', handleMouseMove);
```
[the remark this answers is not in the document]
Comment (TueeNguyen) [reply src/web/src/components/Posts/Post.tsx:408]: I tried adding `mouseenter` here, the `e.target` returns the entire page document always

**Comment by AmasiaNalbandian** [comment]
Hello! Please rebase, as the following commit https://github.com/Seneca-CDOT/telescope/commit/6046273fba4b89ce6281f7a34cafbdec7ef222dc has reorganized our file structure!

File: src/web/src/components/Posts/Post.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (menghif) [reply src/web/src/components/Posts/Post.tsx]: I used `mouseleave` to remove the button as suggested. However, I kept mouse hover to create the button. `mouseenter` does not work when you keep the cursor still and scroll down the page until the cursor is inside the code block.

**Comment by menghif** [comment]
This now shows/hides the button as expected

File: src/web/app/src/components/Posts/Post.tsx [hunk 3]
```diff
@@ -327,6 +337,51 @@ function handleZoom(e: MouseEvent) {
...
+    e.preventDefault();
+    copyCode(e.target);
+  }
+}
+
+function createCopyButton(e: MouseEvent) {
+  if (
+    e.target instanceof HTMLElement &&
+    e.target.tagName === 'CODE' &&
+    e.target.parentElement?.tagName === 'PRE'
```
Comment (TueeNguyen) [inline src/web/app/src/components/Posts/Post.tsx:352]: Could make this a function

File: src/web/app/src/components/Posts/Post.tsx [hunk 4]
```diff
@@ -327,6 +337,51 @@ function handleZoom(e: MouseEvent) {
...
+    elem.parentNode?.removeChild(elem);
+  });
+}
+
+function handleMouseMove(e: MouseEvent) {
+  // if mouse hovers <code></code>, we call createCopyButton(e)
+  if (
+    e.target instanceof HTMLElement &&
+    e.target.tagName === 'CODE' &&
+    e.target.parentElement?.tagName === 'PRE'
```
Comment (TueeNguyen) [inline src/web/app/src/components/Posts/Post.tsx:378]: Same as the previous comment

**Comment by TueeNguyen** [comment]
It's so cool to see you took some inspiration from me 😄

File: src/web/app/src/components/Posts/Post.tsx (hunk 3 above)
Comment (menghif) [reply src/web/app/src/components/Posts/Post.tsx:352]: With all the changes you suggested it complains that `e` is possibly null. I'm not sure how to handle it

Comment (AmasiaNalbandian) [reply src/web/app/src/components/Posts/Post.tsx:352]: `e?.target`

File: src/web/app/src/components/Posts/Post.tsx [hunk 5]
```diff
@@ -360,9 +415,17 @@ const PostComponent = ({ postUrl, currentPost, totalPosts }: Props) => {
 
   // Listen for click events
   useEffect(() => {
-    window.document.addEventListener('click', handleZoom);
+    window.document.addEventListener('click', handleClick);
```
Comment (humphd) [inline src/web/app/src/components/Posts/Post.tsx:418]: I think you need to move `handleClick` into the closure:

```js
useEffect(() => {
  const handleClick = ...;
  window.document.addEventListener('click', handleClick);
  return () => {
    window.document.removeEventListener('click', handleClick); 
  }; 
```

Question: are you doing this at the right level?  Won't this mean we'll have one click event handler on the document for every post vs. one for all posts?  I think this needs to go higher up in the component stack, and you pass make this a controlled component.

**Review by humphd — CHANGES_REQUESTED** [review]
I like this, but I don't think the code is ready.

File: src/web/app/src/components/Posts/Post.tsx (hunk 5 above)
Comment (humphd) [reply src/web/app/src/components/Posts/Post.tsx:418]: Confirm that I'm right about this.  cc @DukeManh

Comment (DukeManh) [reply src/web/app/src/components/Posts/Post.tsx:418]: `handleClick` refers to the same function so I doubt I will make another listener.
I agree that it should be moved up higher in the tree, but the posts content loads individually so I don't know if the listener will apply for the posts that load after

**Comment by DukeManh** [comment]
@menghif I know a bit about `Posts` so I want to collaborate with you on this here https://github.com/menghif/telescope/pull/7

**Review by AmasiaNalbandian — DISMISSED** [review]
I'm good with this. I'm not a fan of the click color being the same color when you are hovering over code (not the copy icon), but that is far from being an issue.

**Comment by DukeManh** [comment]
@AmasiaNalbandian, can you explain more. It's not ready yet, I still want to push something.

**Comment by tpmai22** [comment]
@DukeManh I think what @AmasiaNalbandian mean is that she expect the color will look  slightly different when you click the button then when you hover it.

**Comment by DukeManh** [comment]
Ok, this is cool and ready to be reviewed. I spent some time working on this I wanted the flexibility of React to build the copy button and the copy badge,  I made a `CopyButton` component and render it with `ReactDOM.dom` API into the `pre` tag when the mouse hovers over it. 

Also removed `window.addEventListener('mousemove', ..)` in `useEffect` because the easier way is to add the lisner to the element directly like so `<div onMouseMove={handleMouseMove}/>`.

**Review by humphd — DISMISSED** [review]
Nice.

One thing I notice during testing is that we now have two copy buttons, and they use different colours (new one is dark, old one is light):

Screen Shot 2022-03-28 at 9 08 02 AM

Can we file a follow-up to make them the same?

**Review by AmasiaNalbandian — APPROVED** [review]
This got rid of my previous feedback - I'm also loving the instant feedback that the click has registered and has copied the code block! Nice work team!

**Review by menghif — COMMENTED** [review]
This looks amazing! Thank you @DukeManh for making it great

**Review by humphd — APPROVED** [review]
Looks good, but I'd like to see an issue filed to deal with the difference between how the two copy buttons work.  We should re-use the same component or at least have the styles be the same.

**Review by DukeManh — APPROVED** [review]
Filed #3356 follow-up.

**Pull request merged**